api/ccache_test.go

```diff
@@ -0,0 +1,226 @@
+package api
+
+import (
+	"bytes"
+	"encoding/json"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	"github.com/grafana/metrictank/api/models"
+	"github.com/grafana/metrictank/api/response"
+	"github.com/grafana/metrictank/cluster"
+	"github.com/grafana/metrictank/conf"
+	"github.com/grafana/metrictank/idx/memory"
+	"github.com/grafana/metrictank/mdata"
+	"github.com/grafana/metrictank/mdata/cache"
+	"gopkg.in/raintank/schema.v1"
+)
+
+func newSrv(delSeries, delArchives int, key string) (*Server, *cache.MockCache) {
+	srv, _ := NewServer()
+	srv.RegisterRoutes()
+
+	mdata.SetSingleAgg(conf.Avg, conf.Min, conf.Max)
+	mdata.SetSingleSchema(conf.NewRetentionMT(10, 100, 600, 10, true))
+
+	store := mdata.NewDevnullStore()
+	srv.BindBackendStore(store)
+
+	mockCache := cache.NewMockCache()
+	mockCache.DelMetricRes = cache.CCDelMetricResult{Series: delSeries, Archives: delArchives}
+	metrics := mdata.NewAggMetrics(store, mockCache, false, 0, 0, 0)
+	srv.BindMemoryStore(metrics)
+	srv.BindCache(mockCache)
+
+	metricIndex := memory.New()
+	metricIndex.AddOrUpdate(
+		&schema.MetricData{
+			Id:       key,
+			OrgId:    1,
+			Name:     "test.key",
+			Metric:   "test.key",
+			Interval: 10,
+			Value:    1,
+		},
+		0,
+	)
+	srv.BindMetricIndex(metricIndex)
+	return srv, mockCache
+}
+
+func TestMetricDelete(t *testing.T) {
+	cluster.Init("default", "test", time.Now(), "http", 6060)
+
+	delSeries := 3
+	delArchives := 10
+	testKey := "12345"
+
+	srv, cache := newSrv(delSeries, delArchives, testKey)
+	req, _ := json.Marshal(models.CCacheDelete{
+		Patterns:  []string{"test.*"},
+		OrgId:     1,
+		Propagate: false,
+	})
+
+	ts := httptest.NewServer(srv.Macaron)
+	defer ts.Close()
+
+	res, err := http.Post(ts.URL+"/ccache/delete", "application/json", bytes.NewReader(req))
+	if err != nil {
+		t.Fatalf("There was an error in the request: %s", err)
+	}
+
+	respParsed := models.CCacheDeleteResp{}
+	buf := new(bytes.Buffer)
+	buf.ReadFrom(res.Body)
+	json.Unmarshal(buf.Bytes(), &respParsed)
+
+	if len(cache.DelMetricKeys) != 1 || cache.DelMetricKeys[0] != testKey {
+		t.Fatalf("Expected that key %s has been deleted, but it has not", testKey)
+	}
+
+	if respParsed.DeletedSeries != delSeries || respParsed.DeletedArchives != delArchives {
+		t.Fatalf("Expected %d series and %d archives to get deleted, but got %d and %d", delSeries, delArchives, respParsed.DeletedSeries, respParsed.DeletedArchives)
+	}
+}
+
+func TestMetricDeleteWithErrorInPropagation(t *testing.T) {
+	manager := cluster.InitMock()
+
+	expectedDeletedSeries, expectedDeletedArchives := 0, 0
+
+	// define how many series/archives are getting deleted by peer 0
+	resp := models.CCacheDeleteResp{
+		Peers:           map[string]models.CCacheDeleteResp{"2": {Errors: 1}},
```

Comment:
why are we setting this? this response is the response of a peer that we propagated to, right? so that peer should not propagate

Reply:
Right, removing that line